File: 1.两数之和/1.java
// 破解
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] ans = new int[2];
    	for(int i = 0, len = nums.length; i < len; i++){
    		for(int j = 0, j_len = nums.length; j < j_len; j++){
    			if(i != j && (nums[i] + nums[j] == target)){
    				ans[0] = i;
    				ans[1] = j;
    				break;
    			}
    		}
    	}
        return ans;
    }
}
// 两遍哈希 (dynamic programming)
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement) && map.get(complement) != i) {
                return new int[] { i, map.get(complement) };
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}
// 一遍哈希表 (dynamic programming)
class Solution {
	public int[] twoSum(int[] nums, int target) {
		Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
	}	 
}

File: 1.两数之和/1.cpp
// 暴力破解法
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int i,j;
        for(i=0;i<nums.size()-1;i++)
        {
            for(j=i+1;j<nums.size();j++)
            {
                if(nums[i]+nums[j]==target)
                {
                   return {i,j};
                }
            }
        }
        return {i,j};
    };
};
// 空间换时间(dynamic programming)
// 两遍哈希 
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int,int> a;          //建立hash表存放数组元素
        vector<int> b(2,-1);     //存放结果
        for(int i=0; i < nums.size(); i++){
        	a.insert(map<int,int>::value_type(nums[i],i));
    	}            
        for(int i=0; i < nums.size(); i++){
            if(a.count(target-nums[i]) > 0 && (a[target-nums[i]] != i)){
            //判断是否找到目标元素且目标元素不能是本身
                b[0] = i;
                b[1] = a[target-nums[i]];
                break;
            }
        }
        return b;
    };
};
// 一遍哈希
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int,int> a;           //提供一对一的hash
        vector<int> b(2,-1);      //用来承载结果，初始化一个大小为2，值为-1的容器b
        for(int i = 0; i < nums.size(); i++){
            if(a.count(target-nums[i]) > 0){
                b[0] = a[target-nums[i]];
                b[1] = i;
                break;
            }
            a[nums[i]] = i;       //反过来放入map中，用来获取结果下标
        }
        return b;
    };
};
